Correction of Water Thermal Tank Underflow Problem

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8844 
 - A user running the water thermal tank using a heat pump was running into a NaN crash during a run.  Debugging showed that a very low flow rate (e-300+) was resulting in the NaN problem that was causing the crash.  The fix here was to make sure that the flow rate coming out of a function that was setting the flow rate was limiting things so that if a very small flow rate was being returned that this was reset to zero to avoid these problems.  After the fix, the user's input file runs successfully to completion with no errors.  Pending ci results are showing no differences as anticipated.  A unit test shows that the fix is working and appropriate.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Water Tank Underflow Fix

Correction of the problem.  There were a variety of underflow issues in the water tank model as well as a potential divide by zero.  These were not causing a hard crash but eventually led to a fatal error that stopped EnergyPlus.  The solution stops the problem at the source to void sending incredibly tiny flow values to other routines and leading to problems.

RKStrand committed:
Water Tank Underflow Format Fixes

Formatting issues that were missed in the last commit.

RKStrand committed:
Merge branch 'develop' into 8844WaterThermalTankUnderflowProblem

RKStrand committed:
Water Thermal Tank Fix Revision

One of the changes that was related to this problem but not the ultimate source was causing other issues.  So, this changes was backed out because it is not needed to fix the problem and it was uncertain if this was really necessary.  This fix should work and eliminate some of the problems running the test suite.

RKStrand committed:
Water Thermal Tank Underflow Unit Test

Creation of a unit test that demonstrates the purpose and functionality of the fix to avoid the underflow issue.

RKStrand committed:
Merge branch 'develop' into 8844WaterThermalTankUnderflowProblem

jcyuan2020 commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -10755,6 +10755,9 @@ Real64 WaterThermalTankData::PlantMassFlowRatesFunc(EnergyPlusData &state,
     }
 
+    Real64 const minFlow = 1.0e-20; // Catch underflow problems
+    if (FlowResult < minFlow) FlowResult = 0.0;
```
This is a valid fix and would fix the problem and prevent the flow to take very small values.

jcyuan2020 replied:
(This is beyond this one fix). I think there could be some global small-value defined so that it could also be used in some other places. 

I did a quick search in the code. There are a few places where a SMALL value of 1e-10 defined in some local functions such as in General.hh; but in most locations it is made only local to a function such as safe divisions: 
https://github.com/NREL/EnergyPlus/blob/cbd1174e5e3266057c2e5656fb1a06f9787f6ed2/src/EnergyPlus/General.cc#L909

Maybe making a few of such SMALL values global (with different names and values for different types of physics) might be helpful in the future? @RKStrand @mjwitte @Myoldmopar

rraustad replied:
In DataHVACGlobals:

    Real64 constexpr SmallMassFlow(0.001);
    Real64 constexpr VerySmallMassFlow(1.0E-30);
    Real64 constexpr SmallWaterVolFlow(1.0E-9);

RKStrand replied:
Thanks @rraustad!  I'll change this to the VerySmallMassFlow and recommit.

jcyuan2020 replied:
Thanks @rraustad and @RKStrand !

A side issue that I noticed when running a defect file is about the AIRLOOPHVAC:OUTDOORAIRSYSTEM transition program is that it cuts off the third line for a 3-line version--which calls a little bit treatment on the 3-line case. See #8995.

RKStrand replied:
@jcyuan2020 I have made a commit on this that changes to the existing VerySmallMassFlow constant.  So, this should be good to go.

Thanks for the heads-up on #8995.  I'm going to start working on that as soon as possible so that it can be done today.

RKStrand committed:
Merge branch 'develop' into 8844WaterThermalTankUnderflowProblem

RKStrand committed:
Water Thermal Tank Underflow Using Existing Variable

Based on a review comment, changed the lower limit to an existing variable from DataHVACGlobals.

jcyuan2020 commented:
Thanks @RKStrand ! I also finished running the defect file. It now finishes the simulation without problems---in the develop branch it will crash on div-by-zero when running the debug version. So the current fix is successful. I think it should be good to go in. 

One thing I noticed that the negative water electric reporting (in html) in the defect file still does not go way---it yields the same reporting result as before (by looking at the original helpdesk 15999 issue complaint). But I think this probably should be left to a new investigation regarding reporting on water electricity energy. I originally replied the user to advise that that they might have specified too low a flow rate to the domestic water heater, which could be one of the reasons for the observed abnormalities. The water tank div-by-zero issue is the first problem I encountered that prevented the code to be executed further---and essential that's what been reported in the original issue. 

I think the current fix paved the way for further investigations of the issues reported in the original helpdesk ticket. I will probably transfer the original GitHub issue post to an updated one with the currently known progress---and someone may also take a look at the electricity reporting problem later.

RKStrand commented:
Thanks, @jcyuan2020.  Strange that it still has negative water electric, but I think that is a separate issue.  Yes, please add another issue.  I still have some defects to work on before the code freeze, so I might take that on since that work is still somewhat fresh in my mind.  Just wish the user input file didn't take 2 hours to run on my machine!

RKStrand commented:
@jcyuan2020 Oh...I missed this before.  It seems like there is a unit test issue.  I'll look into that and make another commit as soon as I can.

RKStrand committed:
Water Thermal Tank Unit Test Fix

With the change to the minimum flow parameter from DataHVACGlobals, a changed needed to be made to the unit test as well.  This should now resolved that problem.

RKStrand commented:
@jcyuan2020 Ok, that was a simple problem that was related to using the existing constant to check for minimum flow.  A new commit was made.  Should be good now.

jcyuan2020 reviewed (approved):
This is great! I will issue an update on the original reported issue for the next steps.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 8844WaterThermalTankUnderflowProblem

mjwitte commented:
Haven't seen a no-diff ci result on this branch. Merged in latest develop.

mjwitte commented:
Ok, Mac CI came back with no diffs, but develop has moved again now. Merging.
